File: python/packages/driver/src/arcadedb_driver/facade/stream.py
"""The `query_stream` / `command_stream` pair: `/query` and `/command` as `application/x-ndjson`.

HAND-WRITTEN, unlike `query`/`command`: every generated operation (`execute_query_post.py`,
`execute_command.py`) calls `response.json()` inside its own `_parse_response` - there is no
generated call this could ride even for the plain buffered body, let alone a streamed one. What is
NOT hand-written is the transport: both the sync and async functions below reach for the generated
`Client`'s own pooled `httpx.Client` / `httpx.AsyncClient` via `.stream()` - the same escape hatch
`facade/timeseries.py`'s hand-written `write` already uses - so the base URL, auth headers and
timeout come along for free instead of being reconstructed. A second transport is exactly what
this milestone was designed to avoid; there is only one here to drift.

Three decisions, matching `queryStream`/`commandStream` in `@arcadedb/driver`:

- **Events, not rows.** Each ndjson line is one `NdJsonQueryEvent`, carrying exactly one of
  `record`, `stats` or `error`. Yielding only `record` would discard the `stats` trailer - which
  carries the same `limit`/`returned`/`truncated` `QueryEnvelope` reports at top level for the
  buffered path - and the in-band `error`. A caller who iterates rows and ignores the trailer
  cannot tell a complete result from a truncated one; that is the same hazard `QueryEnvelope`'s own
  docstring names for `query`/`command`. The generated `NdJsonQueryEvent` model is returned
  unaltered rather than normalised into a bespoke type, the same choice `facade/dashboards.py`'s
  `db.promql` and `facade/vector.py` make for their own generated responses.
- **An in-band `error` raises `ArcadeDBError`.** It is a failure the server can only report after
  the 200 status line was already sent - the status line cannot be taken back once the stream has
  started, which is why the contract carries it in band at all instead of as an HTTP status. The
  raised error's status is 200: honest, and no different from what M4 already established for
  `ArcadeDBError`. Raising here means the streaming and buffered paths fail exactly the same way.
- **`query`/`command` are untouched.** They keep returning `QueryEnvelope` and keep sending no
  `Accept` header at all; streaming is a separate pair of methods, not a mode of the existing two.

Two things Python-specific that the TypeScript twin (`facade/stream.ts`) did not have to solve:

1. **Closing the connection on early abandonment.** `httpx`'s `.stream()` is a context manager
   whose `finally` calls `response.close()` - but a generator that `yield`s from inside that block
   does not run that `finally` until the generator itself is finalised, whether by exhaustion, an
   explicit `close()`/`aclose()`, or garbage collection. Nothing extra needs writing for this
   correctly, though: the `with`/`async with` lives directly in the generator body below, so a
   `GeneratorExit` thrown into the suspended `yield` (which is what `.close()`/`.aclose()` and
   `for ... break` both eventually do) propagates out through the block exactly like any other
   exception, and the context manager's `__exit__`/`__aexit__` runs during that unwind - see
   `test_stream.py`'s explicit `.close()`/`.aclose()` tests, modelled on
   `arcadedb-driver-grpc`'s. This is the same class of hazard M6 named for a plain Python `for`
   loop that never closes the iterable it consumes on early exit; the fix here is structural
   rather than an explicit `try`/`finally`, because the iterable being closed is this function's
   own local `with` block, not a caller-supplied one.
2. **Reading a non-2xx body before it can be reported.** A streaming response's body has not been
   read when its status line arrives - accessing `.content` before `.read()`/`.aread()` raises
   `httpx.ResponseNotRead`. `ArcadeDBError` needs that body for the server's error detail, so it is
   read explicitly, exactly once, before `ArcadeDBError` is raised.

`httpx.Response.iter_lines()` / `.aiter_lines()` already buffer a partial line across chunk
boundaries, so unlike `facade/stream.ts`'s hand-rolled decoder this module carries no `remainder`
string of its own. `test_stream.py` asserts the split-line case anyway: it is the property most
likely to regress if this transform is ever swapped for a hand-rolled loop.
"""

from __future__ import annotations

import json
from collections.abc import AsyncGenerator, Generator
from typing import Any
from urllib.parse import quote

import httpx

from .._generated.client import Client
from .._generated.models.nd_json_query_event import NdJsonQueryEvent
from .._generated.types import Unset
from ..errors import REQUEST_ID_HEADER, ArcadeDBError
from .data import SESSION_HEADER, QueryLanguage, build_command_request, build_query_request


def _query_url(database: str) -> str:
    return f"/api/v1/query/{quote(database, safe='')}"


def _command_url(database: str) -> str:
    return f"/api/v1/command/{quote(database, safe='')}"


def _headers(session_id: str | None) -> dict[str, str]:
    headers = {"Accept": "application/x-ndjson"}
    if session_id is not None:
        headers[SESSION_HEADER] = session_id
    return headers


def _parse_event(line: str) -> NdJsonQueryEvent:
    """Parses one ndjson line, raising `ArcadeDBError` if it is an in-band `error` event rather
    than returning it - see the module docstring for why 200 is the right status to carry."""
    event = NdJsonQueryEvent.from_dict(json.loads(line))
    if not isinstance(event.error, Unset):
        message = event.error.message
        detail = "the stream reported an error" if isinstance(message, Unset) else message
        raise ArcadeDBError(200, {"error": detail})
    return event


def _raise_for_status(response: httpx.Response) -> None:
    """Raises `ArcadeDBError` unless the server answered 2xx.

    Reads the body first: a streaming response has not been read yet, and `ArcadeDBError` needs
    the body for the server's detail.
    """
    if not response.is_success:
        response.read()
        raise ArcadeDBError(response.status_code, response.content, response.headers.get(REQUEST_ID_HEADER))


async def _araise_for_status(response: httpx.Response) -> None:
    """The async twin of `_raise_for_status`."""
    if not response.is_success:
        await response.aread()
        raise ArcadeDBError(response.status_code, response.content, response.headers.get(REQUEST_ID_HEADER))


def _stream_events(
    client: Client, url: str, body: dict[str, Any], session_id: str | None
) -> Generator[NdJsonQueryEvent, None, None]:
    with client.get_httpx_client().stream("POST", url, json=body, headers=_headers(session_id)) as response:
        _raise_for_status(response)
        for line in response.iter_lines():
            if not line.strip():
                continue
            yield _parse_event(line)


async def _astream_events(
    client: Client, url: str, body: dict[str, Any], session_id: str | None
) -> AsyncGenerator[NdJsonQueryEvent, None]:
    async with client.get_async_httpx_client().stream("POST", url, json=body, headers=_headers(session_id)) as response:
        await _araise_for_status(response)
        async for line in response.aiter_lines():
            if not line.strip():
                continue
            yield _parse_event(line)


def stream_query(
    client: Client,
    database: str,
    session_id: str | None,
    *,
    language: QueryLanguage,
    command: str,
    params: dict[str, Any] | None = None,
    limit: int | None = None,
) -> Generator[NdJsonQueryEvent, None, None]:
    """Streams `POST /api/v1/query/{database}` as `application/x-ndjson`, one event per line - see
    the module docstring for what an event carries and why an in-band `error` raises."""
    body = build_query_request(language=language, command=command, params=params, limit=limit).to_dict()
    return _stream_events(client, _query_url(database), body, session_id)


def stream_command(
    client: Client,
    database: str,
    session_id: str | None,
    *,
    language: QueryLanguage,
    command: str,
    params: dict[str, Any] | None = None,
) -> Generator[NdJsonQueryEvent, None, None]:
    """Streams `POST /api/v1/command/{database}` as `application/x-ndjson`, one event per line -
    see the module docstring for what an event carries and why an in-band `error` raises."""
    body = build_command_request(language=language, command=command, params=params).to_dict()
    return _stream_events(client, _command_url(database), body, session_id)


def astream_query(
    client: Client,
    database: str,
    session_id: str | None,
    *,
    language: QueryLanguage,
    command: str,
    params: dict[str, Any] | None = None,
    limit: int | None = None,
) -> AsyncGenerator[NdJsonQueryEvent, None]:
    """The async twin of `stream_query`."""
    body = build_query_request(language=language, command=command, params=params, limit=limit).to_dict()
    return _astream_events(client, _query_url(database), body, session_id)


def astream_command(
    client: Client,
    database: str,
    session_id: str | None,
    *,
    language: QueryLanguage,
    command: str,
    params: dict[str, Any] | None = None,
) -> AsyncGenerator[NdJsonQueryEvent, None]:
    """The async twin of `stream_command`."""
    body = build_command_request(language=language, command=command, params=params).to_dict()
    return _astream_events(client, _command_url(database), body, session_id)

File: typescript/packages/driver/src/facade/stream.ts
import type { Client } from "openapi-fetch";
import type { components, paths } from "../generated/schema.js";
import { ArcadeDBError } from "../errors.js";
import { decodeNdJson } from "../internal/ndjson.js";
import { buildCommandBody, buildQueryBody, sessionHeader } from "../internal/request-body.js";
import type { CommandOptions, QueryOptions } from "./data.js";

/** The unwrapped openapi-fetch client, typed against ArcadeDB's OpenAPI schema. */
type RawClient = Client<paths>;

export type NdJsonQueryEvent = components["schemas"]["NdJsonQueryEvent"];

/**
 * Turns one raw `NdJsonQueryEvent` stream into the public generator's contract: an in-band `error`
 * event (D-M7-3) raises `ArcadeDBError` instead of being yielded, after every event that arrived
 * before it has already been handed to the caller. This is a failure the server could only report
 * after the 200 status line was already sent - the status cannot be taken back once the stream has
 * started, so the contract puts it in band instead. Raising it here means the streaming and
 * buffered paths fail the same way: both end in a thrown `ArcadeDBError`, never a returned one.
 */
async function* raiseOnErrorEvent(events: AsyncGenerator<NdJsonQueryEvent>): AsyncGenerator<NdJsonQueryEvent> {
  for await (const event of events) {
    if (event.error !== undefined) {
      throw new ArcadeDBError(200, { error: event.error.message ?? "the stream reported an error" });
    }
    yield event;
  }
}

/**
 * D-M7-1: this module rides the existing generated client - `client.POST(..., { parseAs: "stream" })`
 * - rather than calling `fetch` directly or standing up a second client. `parseAs: "stream"` hands
 * back `response.body` (a `ReadableStream<Uint8Array>`) unparsed instead of JSON-decoding it, but
 * the request that produces that response still goes through the same base URL and the same auth
 * middleware every other call on this client goes through, and openapi-fetch still reads a non-ok
 * response as text/JSON independently of `parseAs` - so the existing `ArcadeDBError` mapping keeps
 * working here too. A second transport was exactly what this milestone risked needing before
 * `parseAs: "stream"` was found to make that unnecessary: hand-writing a second client able to
 * reapply auth and reproduce the error mapping would have doubled the surface this repository has
 * to keep in step with the server, for a feature that only needed a different way to read one
 * response body.
 *
 * What IS hand-written is `decodeNdJson`: the transform from raw bytes to `NdJsonQueryEvent`s. See
 * its doc comment for the two chunk-boundary defects that transform must not have.
 */
async function* streamEvents(
  client: RawClient,
  path: "/api/v1/query/{database}" | "/api/v1/command/{database}",
  database: string,
  sessionId: string | undefined,
  body: { command: string; language: string; params?: Record<string, never>; limit?: number },
): AsyncGenerator<NdJsonQueryEvent> {
  const { data, error, response } = await client.POST(path, {
    params: { path: { database }, header: sessionHeader(sessionId) },
    body,
    headers: { Accept: "application/x-ndjson" },
    parseAs: "stream",
  });
  if (!response.ok) {
    throw ArcadeDBError.fromResponse(response, error);
  }
  // `data` is `response.body`: `null` when the runtime reports no body (e.g. a genuinely empty
  // response), and typed as possibly `undefined` only because it shares a discriminated union
  // with the error branch already ruled out above by the `response.ok` check.
  if (data === null || data === undefined) {
    return;
  }
  yield* raiseOnErrorEvent(decodeNdJson<NdJsonQueryEvent>(data));
}

/** Streams `POST /api/v1/query/{database}` as `application/x-ndjson`, yielding one event per line. */
export async function* queryStream(
  client: RawClient,
  database: string,
  sessionId: string | undefined,
  opts: QueryOptions,
): AsyncGenerator<NdJsonQueryEvent> {
  yield* streamEvents(client, "/api/v1/query/{database}", database, sessionId, buildQueryBody(opts));
}

/**
 * Streams `POST /api/v1/command/{database}` as `application/x-ndjson`, yielding one event per
 * line. Only a read-only statement can stream; see `ArcadeDBDatabase.commandStream`'s doc comment
 * for why a mutating one is refused.
 */
export async function* commandStream(
  client: RawClient,
  database: string,
  sessionId: string | undefined,
  opts: CommandOptions,
): AsyncGenerator<NdJsonQueryEvent> {
  yield* streamEvents(client, "/api/v1/command/{database}", database, sessionId, buildCommandBody(opts));
}
